pos_order_to_sale_order/static/src/js/pos_order_to_sale_order.js

```diff
@@ -0,0 +1,137 @@
+
+openerp.pos_order_to_sale_order = function(instance, local) {
+
+    "use strict";
+
+    var module = instance.point_of_sale;
+    var _t = instance.web._t;
+
+    /*************************************************************************
+        New Widget CreateSaleOrderButtonWidget:
+            * On click, check if there is a customer defined,
+              ask confirmation call server to create sale order, and delete
+              the current order. 
+    */
+    module.CreateSaleOrderButtonWidget = module.PosBaseWidget.extend({
+        template: 'CreateSaleOrderButtonWidget',
+
+        init: function(parent, options){
+            this._super(parent, options);
+            this.display_text = _t("Create Draft Order");
+            if (this.pos.config.iface_create_sale_order_action == 'confirmed_order'){
+                this.display_text = _t("Create Confirmed Order");
+            }
+            if (this.pos.config.iface_create_sale_order_action == 'delivered_picking'){
+                this.display_text = _t("Create Delivered Picking");
+            }
+        },
+
+        renderElement: function() {
+            var self = this;
+            this._super();
+            this.$el.click(function(){
+                var current_order = self.pos.get('selectedOrder');
+                // Prevent empty delivery order
+                if (current_order.get('orderLines').length == 0){
+                    self.pos_widget.screen_selector.show_popup('error',{
+                        'message': _t('Empty Order'),
+                        'comment': _t('There must be at least one product in your order to create Sale Order.'),
+                    });
+                    return;
+                }
+                // Check Customer
+                if (current_order.get('client') ===  null){
+                    self.pos_widget.screen_selector.show_popup('error',{
+                        'message': _t('No customer defined'),
+                        'comment': _t('You should select a customer in order to create a Sale Order. Please select one by clicking the order tab.'),
+                    });
+                    return;
+                }
+                self.pos.pos_widget.screen_selector.show_popup('confirm', {
+                    message: _t('Create Sale Order and discard the current PoS Order ?'),
+                    comment: _t("This operation will permanently destroy the current PoS Order and create a Sale Order, based on the current order lines. You'll have to create later an invoice.\n Note if you had manually changed unit prices for some products, this changes will not been taken into account in the sale order, and should be done manually on the invoice again."),
+                    confirm: function(){
+                        var SaleOrderModel = new instance.web.Model('sale.order');
+                        SaleOrderModel.call('create_order_from_pos', [self.prepare_create_sale_order(current_order)]
```

Comment:
Not sure to understand. renderElement doesn't return a deferred (in the native Odoo). So it will not be interpreted and managed ?